src/native_background.ts

```diff
@@ -282,6 +282,50 @@ export async function getenv(variable: string) {
     return (await sendNativeMsg("env", { var: variable })).content
 }
 
+/** Calls an external program, to either set or get the content of the X selection.
+ *  When setting the selection or if getting it failed, will return an empty string.
+ **/
+export async function clipboard(
+    action: "set" | "get",
+    str: string,
+): Promise<string> {
+    let clipcmd = await config.get("externalclipboardcmd")
+    if (clipcmd == "auto") clipcmd = await firstinpath(["xsel", "xclip"])
+
+    if (clipcmd === undefined) {
+        logger.info("Couldn't find an external clipboard executable")
+        return ""
+    }
+
+    if (action == "get") {
+        let result = await run(clipcmd + " -o")
+        if (result.code != 0) {
+            logger.info(
+                `External command failed with code ${result.code}: ${clipcmd}`,
+            )
+            return ""
+        }
+        return result.content
+    }
+
+    // We're going to need to insert str, which we can't trust, in the clipcmd
+    // In order to do this safely we'll use here documents:
+    // http://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_07_04
+
+    // Find a delimiter that isn't in str
+    let heredoc = "TRIDACTYL"
+    while (str.search(heredoc) != -1) heredoc += Math.round(Math.random() * 10)
+
+    // Use delimiter to insert str into clipcmd's stdin
+    clipcmd = `${clipcmd} -i <<'${heredoc}'\n${str}\n${heredoc}\n`
```

Comment:
@cmcaine please give your opinion on whether this is enough sanitisation. It looks plausible to me.

Reply:
Looks good to me.